packages/admin-panel-server/src/viz-builder/extractors/DashboardVisualisationExtractor.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { yup } from '@tupaia/utils';
+
+import { DashboardVisualisationObject } from '../types';
+import { omitBy, isNil } from 'lodash';
+
+export abstract class DashboardVisualisationExtractor {
```

Comment:
This abstract class is for extracting the generic visualisation presentation to what we need (ie: to dashboard_item + report). I was thinking that for we can extend the class to fit our need. Some examples are:
- When we save a draft visualisation version, we don't need to validate everything in dashboard_item and report
- When we publish a visualisation, we will need to fully validate everything in dashboard_item and report
- When we fetch preview data, we only need to extract the report/data part.
- When we implement step by step play, we also need to extract only up to a certain steps,...

Reply:
Hmmm you do make a good point... I think step-by-step play was something I wasn't considering when I made my suggestion particularly... I can see how pushing all the functionality into a single class could have issues down the line.

Generally in these kinds of situations I try to favor composition over inheritance. For instance, perhaps if we were to separate the validation logic from the extraction logic, we may be able to have an easier time mixing and matching them as is needed...

Regardless I'm really happy to let you make the call on this one for the following reasons:
- You seem to have a pretty solid grasp on the various different edge cases
- I definitely don't have a solid idea of how this should come together
- We are just start out on this project, so refactors will inevitably come regardless as requirements make themselves clear
- I've already held this PR up long enough

Reply:
Sorry for holding up this PR for a long time. I've spent a lot of time trying to get it right in the first place but honestly don't have a solid idea at this early stage as well (I've tried some different ways but thought they were overengineered/clunky). So I went with the simpler approach which is to have one `DashboardVisualisation` object and let it accept different validators to mix and match (as I think validators are the main difference between different endpoints). Plus, I think it's the structure that is easiest to be refactored later when we implement more extracting cases